contracts/token/ERC721/ERC721.sol

```diff
@@ -330,9 +330,32 @@ contract ERC721 is Context, ERC165, IERC721 {
         if (!to.isContract()) {
             return true;
         }
-
-        bytes4 retval = IERC721Receiver(to).onERC721Received(_msgSender(), from, tokenId, _data);
-        return (retval == _ERC721_RECEIVED);
+        bytes memory payload = abi.encodeWithSelector(
+            IERC721Receiver(to).onERC721Received.selector,
+            "onERC721Received(address,address,uint256,bytes)",
+            msg.sender,
+            from,
+            tokenId,
+            _data
+        );
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory returndata) = to.call(payload);
+        if (!success) {
+            if (returndata.length > 0) {
+                uint memOffset;
+                // solhint-disable-next-line no-inline-assembly
+                assembly {
```

Comment:
Can you explain a little bit what you tried to do in this assembly block? I'm not sure it's correct.

Reply:
I'm also not sure. I tried to find some example code to follow based on [your comment here](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/1943#discussion_r336142272). How should I fix this part?

Reply:
Take a look at this assembly code that used to be in `ERC165Checker`:

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/b33b906b6e0ce641b26182589973e723e5131edd/contracts/introspection/ERC165Checker.sol#L103-L104

`encodedParams` is a byte array there. Inside the assembly block, `encodedParams` is a variable that refers to the memory location of the byte array. The length of the array is stored in the first 32 bytes at that memory location, and right after that are the array contents. In this PR you're working with the array `returndata`, and we want to ignore the first 4 bytes of content, so we should read the length and decrease it by 4, then take the memory location and add 4 to it, then store the new length at the new memory location, then that is the new byte array that we want.

Reply:
Hi @frangio can you take a look now and let me know if it does what you described? Thanks.

Reply:
Sorry, I've just realized I was guiding you wrong. If we use assembly `revert` we shouldn't need to do any memory manipulation. The following should be enough.

```solidity
assembly {
    let returndata_size := mload(returndata)
    revert(returndata, add(0x20, returndata_size))
}
```